Source/DafnyCore/Resolver.cs

```diff
@@ -9944,43 +9944,109 @@ public CheckDividedConstructorInit_Visitor(Resolver resolver)
+          } else {
+            return false;
+          }
+        }
+        if (asIDT.EqualitySupport == IndDatatypeDecl.ES.Never) {
+          return true;
+        } else if (asIDT.EqualitySupport == IndDatatypeDecl.ES.ConsultTypeArguments) {
+          return false;
+        } else {
+          if (!scc.Contains(asIDT) && CheckIfEqualityIsDefinitelyNotSupported(asIDT, dependencies)) {
+            return true;
+          }
+        }
+      }
+
+      // For synonym types, we need to continue the check on the supporting type
+      var asSyn = type.AsTypeSynonym;
+      if (asSyn != null) {
+        if (BasicCheckIfEqualityIsDefinitelyNotSupported(asSyn.Rhs, dependencies, scc, M)) {
+          return true;
+        }
+      }
+
+      return false;
+
+    }
 
+    bool CheckIfEqualityIsDefinitelyNotSupported(IndDatatypeDecl dt, Graph<IndDatatypeDecl/*!*/>/*!*/ dependencies) {
       var scc = dependencies.GetSCC(dt);
 
+      // If a constructor is ghost, the datatype cannot support equality
       foreach (var ctor in dt.Ctors) {
-
         if (ctor.IsGhost) {
           return true;
         }
 
+        // If a constructor parameter is ghost, the datatype cannot support equality
         foreach (var arg in ctor.Formals) {
-          var anotherIndDt = arg.Type.AsIndDatatype;
-          if (arg.IsGhost ||
-              (anotherIndDt != null && anotherIndDt.EqualitySupport == IndDatatypeDecl.ES.Never) ||
-              arg.Type.IsCoDatatype ||
-              arg.Type.IsArrowType ||
-              (arg.Type.IsOpaqueType && !arg.Type.SupportsEquality)) {
+          var type = arg.Type;
+          if (arg.IsGhost) {
             return true;
           }
 
-          if (anotherIndDt != null && !scc.Contains(anotherIndDt)) {
-            CheckIfEqualityIsDefinitelyNotSupported(anotherIndDt, dependencies);
+          // First, we check the type constructor
+          if (BasicCheckIfEqualityIsDefinitelyNotSupported(type, dependencies, scc, dt.EnclosingModuleDefinition)) {
+            return true;
           }
 
-          foreach (var type in arg.Type.TypeArgs) {
-            var anotherIndDt_arg = type.AsIndDatatype;
-            if ((anotherIndDt_arg != null && anotherIndDt_arg.EqualitySupport == IndDatatypeDecl.ES.Never) ||
-                type.IsCoDatatype ||
-                type.IsArrowType ||
-                (type.IsOpaqueType && !arg.Type.SupportsEquality)) {
-              return true;
-            }
+          // Second, we check all the parameters of the type constructor
+          foreach (var typea in arg.Type.TypeArgs) {
 
-            if (anotherIndDt_arg != null && !scc.Contains(anotherIndDt_arg)) {
-              CheckIfEqualityIsDefinitelyNotSupported(anotherIndDt_arg, dependencies);
+            // There is one subtlety: sometimes, we do not need to check the arguments
+            // It is for example the case for parameters that are not used, or class and trait types
+            var red = type.AsRedirectingType;
```

Comment:
What does `AsRedirectingType` do?

Reply:
When a constructor has a formal of type A<B,C,D>, for some type A<E,F,G> we need to process both the arguments (B, C, D) and the parameters (E, F, G). AsRedirectingType is the way I have found to access the definition of type A to obtain information on its parameters. If there's a better way to do that, I would like to know.